SoloProjects/clc80/Dashboard/Dashboard/ViewController.swift

```diff
@@ -0,0 +1,70 @@
+//
+//  ViewController.swift
+//  Dashboard
+//
+//  Created by Claudia Maciel on 12/13/20.
+//
+
+import UIKit
+
+class ViewController: UIViewController {
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        let dashboardCard = DashboardCard()
+        view.addSubview(dashboardCard)
+        
+        let image1 = UIImageView(image: #imageLiteral(resourceName: "image1"))
+        let image2 = UIImageView(image: #imageLiteral(resourceName: "image2"))
+        let title = UILabel()
+        let subTitle = UILabel()
+        
+        dashboardCard.addSubview(image1)
+        dashboardCard.addSubview(image2)
+        dashboardCard.addSubview(title)
+        dashboardCard.addSubview(subTitle)
+        
+        dashboardCard.translatesAutoresizingMaskIntoConstraints = false
+        image1.translatesAutoresizingMaskIntoConstraints = false
+        image2.translatesAutoresizingMaskIntoConstraints = false
+        title.translatesAutoresizingMaskIntoConstraints = false
+        subTitle.translatesAutoresizingMaskIntoConstraints = false
+        
+        view.backgroundColor = .lightGray
+     
+        title.font = UIFont.preferredFont(forTextStyle: .title1)
+        subTitle.font = UIFont.preferredFont(forTextStyle: .subheadline)
+        title.text = "$1200"
+        subTitle.text = "45% This week"
+        subTitle.textColor = .gray
+
+ 
+        
+        NSLayoutConstraint.activate([
+            dashboardCard.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: 8),
+            dashboardCard.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 20),
+            dashboardCard.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor, constant: 8),
+            
+            image1.leadingAnchor.constraint(equalTo: dashboardCard.leadingAnchor, constant: 20),
+            image1.topAnchor.constraint(equalTo: dashboardCard.topAnchor, constant: 20),
+            image1.heightAnchor.constraint(equalToConstant: 50),
+            image1.widthAnchor.constraint(equalToConstant: 50),
+            
+            title.leadingAnchor.constraint(equalTo: image1.leadingAnchor),
+            title.topAnchor.constraint(equalTo: image1.bottomAnchor, constant: 20),
+            
+            image2.bottomAnchor.constraint(equalTo: title.centerYAnchor),
+            image2.trailingAnchor.constraint(equalTo: dashboardCard.trailingAnchor, constant: -50),
+            image2.widthAnchor.constraint(equalToConstant: 75),
+            image2.heightAnchor.constraint(equalToConstant: 50),
+            
+            subTitle.leadingAnchor.constraint(equalTo: image1.leadingAnchor),
+            subTitle.topAnchor.constraint(equalTo: title.bottomAnchor, constant: 8),
+        ])
```

Comment:
All this setup should be in `DashboardCard`, part of its private implementation details. The images and labels should be properties of the card.

The `DashboardCard` should expose a public API for providing images and text that can be applied to its subviews. There are two ways of going about it:

1. The properties for images and labels are `private`, and `DashboardCard` provides a method like`func configure(with model: DashboardCard.Model)`, which would take a `Model` struct with UIImage and String properties that `DashbaordCard` can decorate itself with.

2. The properties for images and labels are not private, and the object using the card (in this case, the view controller) sets their properties: `card.someImageView.image = blah`, `card.someLabel.text = blah`, etc.

There are arguments for and against both options. What do you think? Which one would you argue for, and why?

Reply:
I went with the first option because as a developer I find it easier to fill in the preset items that are required when initializing the component. So in my DashboardCard file should I make the properties private? 

I was trying to figure out how to change the background color of the Card itself I tried it in both the ViewController and the DashboardCard and neither is making the background color a different color. What am I missing? I'm sure it's something simple that I just can't see.

Reply:
Hi. I need to clarify what I meant by a "configure" method. What you have in your updated code isn't what I was recommended. When you're writing a custom view, there are — very, very roughly speaking — two kinds of configuration:

1. **Initial setup done once.** - This is stuff like adding subviews, establishing AutoLayout constraints, choosing fonts and colors, etc. These are things that a view would set up all by itself, either during `init` or during `awakeFromNib`, etc. What's important to note here is that this is setup that is common to any content the view might display. For example, if you were writing a Twitter client, the layout code in a tweet cell that puts the avatar *over here* and the user name *over there* is something you would set once, even though that tweet cell might get re-used many times to display many different tweets.

2. **Updating the view to display new content.** - This is stuff that happens once, twice, or a hundred times during the lifetime of the view, such as: setting the `image` property on a UIImageView, setting the `text` of a label, etc. In my example of the tweet cell, this is setting the `image` property on the avatar image view, or the `text` property of a user name label. This happens frequently.

So with your Dashboard card, you also have two kinds of setup. Setup Number 1 is what you currently have implemented in `private func configure()`. There's also additional setup needed to apply a particular image or text to the DashboardCard's subviews. Currently you are doing this in `init`, but that's discouraged. It's better to design the API of your view to be initialized once and re-used as needed.

Here's a stubbed-in version explaining what I mean:

```swift
class DashboardCard: UIView {

    struct Model {
        let image1: UIImage
        let image2: UIImage
        let title: String
        let subtitle: String
    }

    private let imageView1 = UIImageView()
    private let imageView2 = UIImageView()
    private let titleLabel = UILabel()
    private let subtitleLabel = UILabel()

    override init(frame: CGRect) {
        super.init(frame: frame)
        initialSetup()
    }

    func configure(with model: Model) {
        imageView1.image = model.image1
        imageView2.image = model.image2
        titleLabel.text = model.title
        subtitleLabel.text = model.subtitle
    }

    private func initialSetup() {
        // add subviews
        // set up layout constraints
        // set background colors, etc.
    }

}
```

What this setup allows you to do is make all the details that **only** DashboardCard should need to know about as `private` matters. It lets another object that needs to use a DashboardCard just do the following:

```swift
class SomeViewControllerSomewhere: UIViewController {

    @IBOutlet private var dashboardCard: DashboardCard!

    func didRefreshAndGetSomeDataOrWhatever(data: SomeModelType) {
        let cardModel = DashboardCard.Model(image1: data.i1, image2: data.i2, title: data.t, subtitle: data.s)
        dashboardCard.configure(with: cardModel)
    }

}

```

What you can see here is that `SomeViewControllerSomewhere` doesn't need to know anything about the private details of DashboardCard or it's subviews. All it needs, other than adding a `DashboardCard` to it's view hierarchy, is to call `configure(with: DashboardCard.Model)` on the card, passing the model values it wants the card to present. Don't pass actual views (don't pass image views or labels). Instead, pass _values_: images, strings, bools, that sort of thing.